DuckDuckGo/NetworkProtectionDNSSettingsViewModel.swift

```diff
@@ -0,0 +1,79 @@
+//
+//  NetworkProtectionDNSSettingsViewModel.swift
+//  DuckDuckGo
+//
+//  Copyright © 2024 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import Combine
+import NetworkProtection
+import Core
+
+final class NetworkProtectionDNSSettingsViewModel: ObservableObject {
+    private let settings: VPNSettings
+    private var cancellables: Set<AnyCancellable> = []
+
+    @Published public var dnsSettings: NetworkProtectionDNSSettings = .default
+
+    @Published public var isCustomDNSSelected = false
+    
+    @Published public var customDNSServers = ""
+
+    @Published public var isApplyButtonEnabled = false
+
+    init(settings: VPNSettings) {
+        self.settings = settings
+
+        settings.dnsSettingsPublisher
+            .receive(on: DispatchQueue.main)
+            .assign(to: \.dnsSettings, onWeaklyHeld: self)
+            .store(in: &cancellables)
+
+        isCustomDNSSelected = settings.dnsSettings.usesCustomDNS
+        customDNSServers = settings.dnsSettings.dnsServersText
+    }
+
+    func toggleDNSSettings() {
+        isCustomDNSSelected.toggle()
+    }
+
+    func applyDNSSettings() {
+        if isCustomDNSSelected {
+            settings.dnsSettings = .custom([customDNSServers])
+            DailyPixel.fireDailyAndCount(pixel: .networkProtectionDNSUpdateCustom)
+        } else {
+            settings.dnsSettings = .default
+            DailyPixel.fireDailyAndCount(pixel: .networkProtectionDNSUpdateDefault)
+        }
+    }
+
+    func updateApplyButtonState() {
+        if isCustomDNSSelected {
+            isApplyButtonEnabled = !customDNSServers.isEmpty && customDNSServers.isValidIpv4Host
```

Comment:
`isValidIpv4Host` will use leading zeroes if you don't provide a full address, i.e. if you enter the string `1` it will allow you to apply it and expand it to `0.0.0.1`. This expansion behaviour is fine since it expands to a legit IP, but it may be confusing for users since we don't show the expanded value - a knowledgable user would likely guess correctly, but I doubt all users would.

When the "Apply" button is tapped, we could use the IP string to initialize an `IPv4Address` then get the full string being used out of that, and save _that_ to storage. In the case of the string `1`, you can call `String(reflecting: address)` to get back `0.0.0.1`.

Reply:
One other kind of strange thing that the `IPv4Address` string initializer does is take a subset of the string you provide it to extract the IP. For example, you can use `IPv4Address("123412341234123412341234123412341234")` and get back an IP address of `37.12.1.242`, but in the UI we will show you that full obviously-incorrect string.

Reply:
Interesting. I updated it so that what you see is what you get at least.